webauthn-server-core/src/main/java/com/yubico/webauthn/OriginMatcher.java

```diff
@@ -16,38 +16,56 @@ static boolean isAllowed(
         boolean allowPort,
         boolean allowSubdomain
     ) {
+        log.trace("isAllowed({}, {}, {}, {})", origin, allowedOrigins, allowPort, allowSubdomain);
+
+        final URL originUrl;
+        {
+            URL originUrl1;
+            try {
+                originUrl1 = new URL(origin);
+            } catch (MalformedURLException e) {
+                log.debug("Origin in client data is an invalid URL; will only match exactly: {}", origin);
+                originUrl1 = null;
+            }
+            originUrl = originUrl1;
+        }
+
         return allowedOrigins.stream().anyMatch(allowedOriginString -> {
             if (allowedOriginString.equals(origin)) {
+                log.debug("Exact match: {} == {}", origin, allowedOriginString);
                 return true;
-            } else {
+            } else if (originUrl != null && (allowPort || allowSubdomain)) {
                 final URL allowedOrigin;
                 try {
                     allowedOrigin = new URL(allowedOriginString);
                 } catch (MalformedURLException e) {
-                    log.error("Allowed origin is an invalid URL: {}", allowedOriginString);
+                    log.error("Allowed origin is an invalid URL; skipped for port/subdomain matching: {}", allowedOriginString);
                     return false;
                 }
 
-                final URL originUrl;
-                try {
-                    originUrl = new URL(origin);
-                } catch (MalformedURLException e) {
-                    log.debug("Invalid origin in client data: {}", origin);
-                    return false;
-                }
+                final boolean portAccepted;
+                final boolean domainAccepted;
 
-                if (!allowPort && originUrl.getPort() != allowedOrigin.getPort()) {
-                    return false;
+                if (allowPort) {
+                    portAccepted = true;
+                } else {
+                    portAccepted = originUrl.getPort() == allowedOrigin.getPort();
                 }
 
                 final String allowedDomain = allowedOrigin.getHost();
                 final String originDomain = originUrl.getHost();
 
                 if (allowSubdomain) {
-                    return originDomain.endsWith("." + allowedDomain);
+                    domainAccepted = originDomain.equals(allowedDomain) || originDomain.endsWith("." + allowedDomain);
                 } else {
-                    return originDomain.equals(allowedDomain);
+                    domainAccepted = originDomain.equals(allowedDomain);
                 }
+
+                log.debug("portAccepted: {}, domainAccepted: {}", portAccepted, domainAccepted);
+                return portAccepted && domainAccepted;
+            } else {
+                log.debug("No match: {} != {}", origin, allowedOriginString);
+                return false;
             }
         });
     }
```

Comment:
In general, it is considered good practice to have the method arguments declared as final since they are references that will continue to live after this method exits and so we want to make sure that we do not accidentally change them.

On the other hand, do the local variables `originUrl`, `portAccepted` and `domainAccepted` really need to be final if that effects the code readability? After all, they will only exist within this method.

Reply:
As far as I know, declaring reference-type parameters as `final` only makes the _reference_ immutable, but the value it points to is still mutable. This did made me notice that I hadn't made immutable the value that gets passed in here as `origins`, so thanks for pointing that out!

`originUrl` needs to be final (or effectively final) because it's referenced in the `anyMatch` lambda. For some reason the static analyzer doesn't accept this:

```
        final URL originUrl;
        try {
            originUrl = new URL(origin);
        } catch (MalformedURLException e) {
            originUrl = null;
        }
```

saying the variable might be assigned twice even though that should be impossible.

`portAccepted` and `domainAccepted` don't strictly need to be final, but I think it helps communicate that they hold results rather than variables. I considered extracting functions for computing them but figured that would probably be overkill, but maybe it's worth it.

Reply:
Ok :-)